libvast/src/meta_index.cpp

```diff
@@ -13,67 +13,194 @@
 
 #include "vast/meta_index.hpp"
 
-#include "vast/event.hpp"
-#include "vast/expression_visitors.hpp"
+#include "vast/expression.hpp"
 #include "vast/logger.hpp"
 #include "vast/table_index.hpp"
 #include "vast/table_slice.hpp"
+#include "vast/time.hpp"
 
-namespace vast {
+#include "vast/detail/overload.hpp"
+#include "vast/detail/set_operations.hpp"
+#include "vast/detail/string.hpp"
 
-// -- member types -------------------------------------------------------------
+namespace vast {
 
-meta_index::interval::interval()
-  : from(timestamp::max()),
-    to(timestamp::min()) {
-  // nop
+meta_index::meta_index() : make_synopsis_{make_synopsis} {
 }
 
-meta_index::interval::interval(timestamp first, timestamp last)
-  : from(first),
-    to(last) {
-  // nop
+void meta_index::add(const uuid& partition, const table_slice& slice) {
+  auto& part_synopsis = partition_synopses_[partition];
+  auto& layout = slice.layout();
+  if (blacklisted_layouts_.count(layout) == 1)
+    return;
+  auto i = part_synopsis.find(layout);
+  table_synopsis* table_syn;
+  if (i != part_synopsis.end()) {
+    table_syn = &i->second;
+  } else {
+    // Create new synopses for a layout we haven't seen before.
+    i = part_synopsis.emplace(layout, table_synopsis{}).first;
+    table_syn = &i->second;
+    for (auto& field : layout.fields)
+      if (i->second.emplace_back(make_synopsis_(field.type)))
```

Comment:
It's not obvious to me what this `if()` is checking. `emplace_back` always succeeds and returns a reference to the added element. Is this a null-pointer check? If so, I think a `!= nullptr` check would make this much more readable.

Reply:
Yep, that is a `nullptr` check. I've made it explicit now.